src/redis-benchmark.c

```diff
@@ -1771,6 +1823,162 @@ int main(int argc, char **argv) {
 
     tag = "";
 
+#if BENCH_CRC_AVAILABLE
+
+#define RPOLY UINT64_C(0x95ac9329ac4bc9b5)
+#define INIT_SIZE UINT64_C(0xffffffffffffffff)
+
+    if (config.crc64_test_size > 0 || config.crc64_combine) do {
+        unsigned char* data = NULL;
+        if (config.crc64_test_size > 0) {
+            data = zmalloc(config.crc64_test_size);
+            if (!data) {
+                fprintf(stderr, "Could not allocate sufficient data to test crc64 "
+                                "%lld\n", config.crc64_test_size);
+                exit(1);
+            }
+        }
+
+/* Will fix later; minimizing CPU used for getting nstime, as shifts and even
+ * *returns* of values are non-zero nanosecond cost in comparison to just *not*
+ * doing those things. So we won't when we need to benchmark crc64 of short
+ * strings, or when we need to benchmark the <100 nanosecond timing of crc
+ * combining. */
+
+        init_nstime();
+
+        struct timespec init_start, init_end;
+        NSTIME(&init_start);
+        /* need to benchmark the combine init first, before we get to crc anything */
+        crc64_combine(
+            UINT64_C(0xdeadbeefdeadbeef),
+            UINT64_C(0xfeebdaedfeebdaed),
+            INIT_SIZE,
+            RPOLY, 64);
+        NSTIME(&init_end);
+
+        crc64_init();
+
+        // We want to hash about 1 gig of data in total, looped, to get a good
+        // idea of our performance.
+        uint64_t expect = 0;
+
+        if (config.crc64_test_size > 0) {
+            genBenchmarkRandomData((char*)data, config.crc64_test_size);
+            // warm up the cache
+            set_crc64_cutoffs(config.crc64_test_size+1, config.crc64_test_size+1);
+            expect = crc64(0, data, config.crc64_test_size);
+            if (config.csv) printf("algorithm,buffer,performance,crc64_matches\n");
+        } else {
+            // just need some nonzero bytes here
+            genBenchmarkRandomData((char*)(&expect), sizeof(uint64_t));
+        }
+
+        // We have nanoseconds, and we need megs/second from bytes. Great,
+        // just a multiply and a division.
+        // By sampling over a range of --crc <sizes>, you can determine the
+        // best values for for setting CRC64_[DUAL|TRI]_CUTOFF on your machine.
+        // NOTE: we needed nstime here for handling timing of processing --crc
+        // values < 4k or so, as they can complete in under 1 microsecond on
+        // some platforms.
+
+        // save some copy / paste
+#define bench(WHICH) \
```

Comment:
Not quite clear why these are macros and not functions. I can't piece out anything that seems useful as a macro, and the macro replacements make this code much less readable.

Reply:
Switching to functions.